### ISSUE TITLE
Improve `CONFIG -h` output

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#3769]
Improve `CONFIG -h` output so the list of possible values is indented.

`config -h machine` output (no other settings are affected):

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/b93a5e53-0094-459b-a8cb-fe45180dda3e)

**Comment by Dustman0192** [comment]
So I fiddled with this one a bit too, but the only solution I could come up with was to pad the values with a \n and then 18 spaces: 

<img width="842" height="359" alt="Image" src="https://github.com/user-attachments/assets/9fef2b5d-cfc3-4113-9481-08be5babb0a5" />

I think that solution is rather inelegant. But unless you guys know of a better way, it at least does the trick:

<img width="842" height="359" alt="Image" src="https://github.com/user-attachments/assets/f7534ae7-44b2-40c5-b658-2518a3ebfea4" />

Let me know and I'll spin up a PR and get this fix going.

**Comment by johnnovak** [comment]
@Dustman0192 Sorry, that won't do it 😄 The proper solution is to implement a mini text wrapping engine that adapts the output to the screen width. That's usually 80 chars, but it can be 40 or 132 too.

**Comment by Dustman0192** [comment]
@johnnovak Got it! OK, still learning the ropes of things :) This one will give me something to chew on for a while!

**Comment by johnnovak** [comment]
Yeah so if you look at the screenshot above and compare it with the string in the code, you'll see it's wrapped manually with newline characters, and indented manually with spaces at the start of each line. So it would be better if descriptions like `cga_mono` and `svga_paradise` would be a _single line_ in the string. In other words, only the paragraphs and bullet point items would be separated by newline chars, then the text would be wrapped intelligently to the available width.

I've expanded the scope of this ticket (check the top description now 😄).

**Comment by Dustman0192** [comment]
I accidentally posted a duplicate screenshot above. Here is the screenshot I meant to post:

<img width="1867" height="308" alt="Image" src="https://github.com/user-attachments/assets/5750ca41-763b-4ec9-b821-63a9e1106c3f" />

Is that the kind of wrapping you're looking for? Or would you prefer to have each of the "Possible values" on their own lines? I can do either or and actually did start thinking about a small wrapping engine that could wrap text within a given character length limit (say we have a max 80 char line length, then values that cause the length to go over that limit would be wrapped to the next line.)

**Comment by Dustman0192** [comment]
I did attempt to find a well tested library for doing wrapping and found this one: https://github.com/asap-projects/asap-textwrap. However, it hasn't had any updates for 2 years so may no longer be maintained.

**Comment by shermp** [comment]
Honestly, some sort of text wrapper function/library would be best. Conceptually, it's not that hard:

1. Find possible break points. Either roll own implementation, or do it properly using a library like [libunibreak](https://github.com/adah1972/libunibreak), because it turns out proper line breaking is [rather complex](https://www.unicode.org/reports/tr14/).
2. Find candidate break points taking into account desired line length etc. One wrinkle is handling ANSI escape codes for colour etc. Thankfully we only have to deal with monospace text.
3. Break the lines, and feed them to output.

That asap-textwrap library looks neat, although I don't know how it could handle the ANSI escape codes.

**Comment by johnnovak** [comment]
Okay, so my view: let's not bring in the tanks, rocket launchers, and special forces units when all we want is catch some mice that wandered into the pantry 😆 

Specifically:

- Let's keep it simple—we don't need the complication of Unicode breaking point rules. Breaking only on spaces and dash characters (minus signs) is our sweet spot. That's perfectly sufficient for the job at hand.
- `asap-textwrap` seemed good for about 5 seconds, but yeah, unmaintained, then I had a look at the sources. Way too complicated for what it's doing. Integrating that into our source tree and maintaining it would be a pain. Well, I don't wanna touch it, personally.
- Handling ANSI control sequences is a very good point; the wrapper engine needs to understand to treat those as "zero length", effectively. Which means we need our custom solution.
- We don't even need complicated multi-pass tokenisation, etc. Way overkill. I did a simple text wrap "engine" (more like a function) a while ago that can handle variable-length glyphs (so for TTF fonts, in other words). You can do the whole thing in a single pass for a simple wrapper implementation; you just need to keep track of the current line length in a variable, then only break on hyphens and spaces, and when your running count is over the line length, move the word to the next line. That's all there is to it, in a nutshell.
- This is a simple header-only implementation of a similar thing in C++: https://github.com/catchorg/textflowcpp/blob/master/TextFlow.hpp I'm not saying to necessarily use this as basis, but we could and then hack it further. In any case, I expect this whole text wrapping thing to be well under 1000 lines, ideally around 500 lines or so. If it's significantly more complex, we're waaaaay overengineering it 😄 
- We need to handle the indentation correctly with list items in config descriptions. A simple rule could be if the line starts with two spaces, that means it's a list item. So take the next word after the two spaces as the "item name", skip the following whitspaces, then treat the rest as the description. For example, `  cga_mono:       CGA adapter...`. The engine will take the column where the description `CGA adapter...` starts and wrap the description to the available width, using the start column to know how many spaces it needs to pad the lines with starting from the second line.

So yeah, keep it simple, keep it line-based, single-pass only, no tokenisation, no complications 😄 Caching is fine, I guess. We could just do a single big hashmap, keys are the message keys, values are the wrapped text. Then we just keep the last 1000 entries or so.

We should come up with something that can be used for the help text of DOS commands too, or pretty much anything we wanna print to the console. We just need a few very simple basic rules.

**Comment by johnnovak** [comment]
To clarify, only left alignment (unjustified) would be supported. That's the key to keep it really simple, that's all we need for monospaced text.

**Comment by Dustman0192** [comment]
How about I just put together something simple for now, something we can look at, and then we can iterate over it and improve it from there?

**Comment by johnnovak** [comment]
> How about I just put together something simple for now, something we can look at, and then we can iterate over it and improve it from there?

Sounds good 😄 That's my idea, start small.

**Comment by Dustman0192** [comment]
Alright! You can assign this issue to me if you'd like. I'll work on it :) I'm also learning C++ as I go along. I have an extensive background in C#, and some junior knowledge of C++. So I'm very much learning as I go along!

**Comment by johnnovak** [comment]
> Alright! You can assign this issue to me if you'd like. I'll work on it :) I'm also learning C++ as I go along. I have an extensive background in C#, and some junior knowledge of C++. So I'm very much learning as I go along!

Looking forward to your solution! 😎 Hope you'll have fun learning C++ 😄 Check out our [contributor's guide](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/CONTRIBUTING.md) if you haven't already done so (I might have already sent you a link... sorry, shit memory 😆)

**Comment by Dustman0192** [comment]
Do any of you have any ideas for how we can do the indentation? I was padding with spaces using `std::string({space_count}, ' ')` where space count is the length of "Possible values:" plus some extra spaces., but it feels like there could be a more elegant way.

The space padding is needed to line up the wrapped text in its own column so it continues to show up under the listed options versus showing up all the way to the left under "Possible values:".

**Comment by shermp** [comment]
Just use spaces, it's the simplest, easiest way to do it, and how I would do it.

It's possible to get creative with ANSI escape codes for positioning, but lets not go there.

The way I would do it would have a function that when given text to wrap, and the number of columns, return lines of wrapped text space padded to the number of columns. Making text centre/right aligned then becomes a simple matter of adjusting the padding location if we wanted to add such an enhancement.

**Comment by johnnovak** [comment]
Yeah, just using spaces is the way. We'll use the wrapped output when writing the config with `config -wc` or `config -wcd`, too, so it's best to use spaces as people have their viewers/editors set to all sorts of tab widths.

Implementation-wise, strings are mutable in C++ so you can build them up with `append()`. Just set the rough expected final size with `reserve()` on the result string to minimize reallocs.

`append()` even has a form to repeat a char; check the first example [here](https://en.cppreference.com/w/cpp/string/basic_string/append.html).

**Comment by Dustman0192** [comment]
Got it! OK, since I'm not taking on that drive label issue anymore I can devote my attention to this :)

**Comment by Dustman0192** [comment]
So I've written the text wrapping function in such a way that it will allow for indenting newlines with a variable number of spaces.

The issue now is: how do we determine how many spaces to indent? From what I gather, the "Possible values:" text output via the "config -h machine" command is created via format string here:

https://github.com/dosbox-staging/dosbox-staging/blob/3382ddb39b9ce807f5551d4156d2eb02ade6a50d/src/misc/programs.cpp#L1006C1-L1007C56

Now, it's easy enough to determine the label length from here (it would just be the length of "Possible values:" plus two spaces.) But do translations affect this label? Would the label length need to be dynamically determined based on them? Or is it largely just a fixed value?

**Comment by Dustman0192** [comment]
I'll go ahead and get a PR going so we can at least look at the work I've done so far and coax it as necessary.

**Comment by johnnovak** [comment]
> Now, it's easy enough to determine the label length from here (it would just be the length of "Possible values:" plus two spaces.) But do translations affect this label? Would the label length need to be dynamically determined based on them? Or is it largely just a fixed value?

You can't assume any fixed lengths in any text because of the translations, yeah. You either need to figure things out dynamically from the text alone, or the client code can "help".Let me explain 😄 

We don't _have_ to make wrapping a single pass operation performed only at the entire final concatenated text message to be printed. Perhaphs it simplifies things if we have calls to wrap on the fragments with different params.

E.g., we could call wrap on the fragment `Possible values: abc, def, ghi,... , xyz` with two parameters: `line_length` and `indent_size` set to the length of the "possible values" fragment.

Wrapping on fragments like this can reduce the number of dynamic rules you need to implement, as the code rendering the fragment usually has extra knowledge how the wrapping should be done exactly.

**Comment by Dustman0192** [comment]
I'll think on this. In the mean time, I've submitted a PR so you can take a look at what I have thus far and root out any issues. The indenting doesn't work yet, so I was hoping we could take a look at that and determine how to make it work reliably with translations.
